temporalio/bridge/src/worker.rs

```diff
@@ -79,6 +85,223 @@ pub struct ResourceBasedSlotSupplier {
 }
 
+#[pyclass]
+pub struct SlotReserveCtx {
+    #[pyo3(get)]
+    pub slot_type: String,
+    #[pyo3(get)]
+    pub task_queue: String,
+    #[pyo3(get)]
+    pub worker_identity: String,
+    #[pyo3(get)]
+    pub worker_build_id: String,
+    #[pyo3(get)]
+    pub is_sticky: bool,
+}
+
+impl SlotReserveCtx {
+    fn from_ctx(slot_type: SlotKindType, ctx: &dyn SlotReservationContext) -> Self {
+        SlotReserveCtx {
+            slot_type: match slot_type {
+                SlotKindType::Workflow => "workflow".to_string(),
+                SlotKindType::Activity => "activity".to_string(),
+                SlotKindType::LocalActivity => "local-activity".to_string(),
+            },
+            task_queue: ctx.task_queue().to_string(),
+            worker_identity: ctx.worker_identity().to_string(),
+            worker_build_id: ctx.worker_build_id().to_string(),
+            is_sticky: ctx.is_sticky(),
+        }
+    }
+}
+
+#[pyclass]
+pub struct SlotMarkUsedCtx {
+    #[pyo3(get)]
+    slot_info: PyObject,
+    #[pyo3(get)]
+    permit: PyObject,
+}
+
+// NOTE: this is dumb because we already have the generated proto code, we just can't use
+//   it b/c it's not pyclassable. In theory maybe we could compile-flag enable it in the core
+//   protos crate but... that's a lot for just this. Maybe if there are other use cases.
+
+#[pyclass]
+pub struct WorkflowSlotInfo {
+    #[pyo3(get)]
+    pub workflow_type: String,
+    #[pyo3(get)]
+    pub is_sticky: bool,
+}
+#[pyclass]
+pub struct ActivitySlotInfo {
+    #[pyo3(get)]
+    pub activity_type: String,
+}
+#[pyclass]
+pub struct LocalActivitySlotInfo {
+    #[pyo3(get)]
+    pub activity_type: String,
+}
+
+#[pyclass]
+pub struct SlotReleaseCtx {
+    #[pyo3(get)]
+    slot_info: Option<PyObject>,
+    #[pyo3(get)]
+    permit: PyObject,
+}
+
+fn slot_info_to_py_obj(py: Python<'_>, info: SlotInfo) -> PyObject {
+    match info {
+        SlotInfo::Workflow(w) => WorkflowSlotInfo {
+            workflow_type: w.workflow_type.clone(),
+            is_sticky: w.is_sticky,
+        }
+        .into_py(py),
+        SlotInfo::Activity(a) => ActivitySlotInfo {
+            activity_type: a.activity_type.clone(),
+        }
+        .into_py(py),
+        SlotInfo::LocalActivity(a) => LocalActivitySlotInfo {
+            activity_type: a.activity_type.clone(),
+        }
+        .into_py(py),
+    }
+}
+
+#[pyclass]
+#[derive(Clone)]
+pub struct CustomSlotSupplier {
+    inner: PyObject,
+}
+
+struct CustomSlotSupplierOfType<SK: SlotKind> {
+    inner: PyObject,
+    _phantom: PhantomData<SK>,
+}
+
+#[pymethods]
+impl CustomSlotSupplier {
+    #[new]
+    fn new(inner: PyObject) -> Self {
+        CustomSlotSupplier { inner }
+    }
+}
+
+#[async_trait::async_trait]
+impl<SK: SlotKind + Send + Sync> SlotSupplierTrait for CustomSlotSupplierOfType<SK> {
+    type SlotKind = SK;
+
+    async fn reserve_slot(&self, ctx: &dyn SlotReservationContext) -> SlotSupplierPermit {
```

Comment:
Is this interruptible? I am curious on worker shutdown what happens here. I see `test_blocking_slot_supplier`, I am just curious what the observed behavior is

Reply:
Well it should be happening automatically in the sense that if you drop the Rust future, it will cause this `CancelledError` to get raised: https://docs.rs/pyo3-asyncio/0.20.0/src/pyo3_asyncio/lib.rs.html#656

Reply:
I think that call just instantiates a cancelled error, it is not actually cancelling the Python awaitable

Reply:
This happens now